internal/sshtunneler/authentication.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package sshtunneler
+
+import (
+	"encoding/base64"
+	"time"
+
+	"github.com/juju/clock"
+	"github.com/juju/errors"
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/lestrrat-go/jwx/v2/jwt"
+)
+
+// tunnelAuthentication provides a way of creating
+// and validating passwords for SSH connections
+// from machines to the Juju controller.
+// The password is a JWT with an expiry.
+type tunnelAuthentication struct {
+	sharedSecret []byte
+	jwtAlg       jwa.KeyAlgorithm
+	clock        clock.Clock
+}
+
+func (tAuth *tunnelAuthentication) generatePassword(tunnelID string, expiry time.Time) (string, error) {
+	token, err := jwt.NewBuilder().
+		Issuer(tokenIssuer).
+		Subject(tokenSubject).
+		IssuedAt(tAuth.clock.Now()).
+		Expiration(expiry).
+		Claim(tunnelIDClaim, tunnelID).
+		Build()
+	if err != nil {
+		return "", errors.Annotate(err, "failed to build token")
+	}
+
+	signedToken, err := jwt.Sign(token, jwt.WithKey(tAuth.jwtAlg, tAuth.sharedSecret))
+	if err != nil {
+		return "", errors.Annotate(err, "failed to sign token")
+	}
+
+	return base64.StdEncoding.EncodeToString(signedToken), nil
+}
+
+func (tAuth *tunnelAuthentication) validatePassword(password string) (string, error) {
+	decodedToken, err := base64.StdEncoding.DecodeString(password)
+	if err != nil {
+		return "", errors.Annotate(err, "failed to decode token")
+	}
+
+	token, err := jwt.Parse(decodedToken,
+		jwt.WithKey(tAuth.jwtAlg, tAuth.sharedSecret),
+		jwt.WithClock(tAuth.clock),
+	)
+	if err != nil {
+		return "", errors.Annotate(err, "failed to parse token")
+	}
+
+	tunnelID, ok := token.PrivateClaims()[tunnelIDClaim].(string)
```

Comment:
it's a bit strange that we're returning a tunnel id from a method called `validatePassword` - contradicts your statement above that we're trying not to expose implementation details, but this seems very much like an implementation detail here.. 
perhaps we should be very explicit in our naming and call it a `jwtTunnelAuthenticator` and have `createJWT` and `getClaims` methods. the latter would just return all claims (the only one being `tunnelIDClaim`)

Reply:
This is not so much an implementation detail, but a core part of how the tunneler works. Because the tunneler works on the assumption of SSH tunnels, we know that on one end there is a consumer requesting a tunnel, and the thing providing those tunnels is an SSH server. The SSH server receives these tunnel requests, must authenticate them, then store a reference to the tunnel ID so that after the connection is established it can push the tunnel to the tracker.

Writing this out though, I realise we could avoid having this validatePassword method return a tunnel ID **IF** the sshsession worker sends the tunnelID back to us when it connects. Also doable, would you want me to make that change? We can also do it later, but overall it's not big deal either way.